src/GitHub/index.jsx

```
@@ -0,0 +1,158 @@
+      redirect_uri: redirectUri,
+      response_type: "code",
+    });
+
+    const popup = PopupWindow.open(`${authorizationUrl}?${search}`);
+    this.popup = popup;
+
+    this.onRequest();
+    popup.then(
+      (data) => this.onSuccess(data),
+      (error) => this.onFailure(error)
+    );
+  }
+
+  onRequest() {
+    const { onRequest } = this.props;
+
+    onRequest();
+  }
+
+  async onSuccess(data) {
+    if (!data.code) {
+      return this.onFailure(new Error("'code' not found"));
+    }
+
+    const AuthorizeUrl =
+      `https://cors.snek.at/` +
+      `https://github.com/login/oauth/access_token?code=${data.code}` +
+      `&client_id=${this.props.clientId}` +
+      `&client_secret=${this.props.clientSecret}` +
+      `&redirect_uri=${this.props.redirectUri}` +
+      `&state=${await GuidGenerator()}`;
+
+    /* POST request to get the access token from GitHub */
+    await fetch(AuthorizeUrl, {
+      // prettier-ignore
+      headers: {
+        "Accept": "application/json",
+        "Access-Allow-Credentials": "True",
+        "Access-Control-Allow-Methods": "POST",
+        "Access-Control-Allow-Origin": "*",
+        "Content-Type": "application/json",
+        "Vary": "Origin",
+      },
+      method: "POST",
+    })
+      .then(async (res) => await res.json())
+      .then(async (res) => {
+        const accessToken = res.access_token;
+
+        /* GET request to get the user used for OAuth */
+        await fetch(`https://api.github.com/user`, {
+          // prettier-ignore
+          headers: {
+            "authorization": "Token " + accessToken,
+          }
+        })
+          .then(async (res) => await res.json())
+          .then((res) => {
+            data = { username: res.login, accessToken };
+
+            const { onSuccess } = this.props;
+
+            return onSuccess(data);
+          });
+      });
+  }
+
+  onFailure(error) {
+    const { onRequest } = this.props;
+    console.log(error);
+    onRequest(error);
```

I would just throw an error. Somebody will catch it for sure.